crates/wallet/src/wallet/tx_builder.rs

```diff
@@ -1098,4 +1090,61 @@ mod test {
         builder.fee_rate(FeeRate::from_sat_per_kwu(feerate + 250));
         let _ = builder.finish().unwrap();
     }
+
+    #[test]
+    fn not_duplicated_utxos_in_required_list() {
+        let mut params = TxParams::default();
+        let test_utxos = get_test_utxos();
+        for _ in 0..3 {
+            params.utxos.insert(test_utxos[0].clone());
+        }
+        assert_eq!(
+            vec![test_utxos[0].clone()],
+            params.utxos.into_iter().collect::<Vec<_>>()
+        );
+    }
+
+    #[test]
+    fn not_duplicated_foreign_utxos_with_same_outpoint_but_different_weight() {
+        use crate::test_utils::get_funded_wallet_wpkh;
+        // Use two different wallets to avoid adding local utxos
+        let (wallet1, txid1) = get_funded_wallet_wpkh();
```

Comment:
You may want to use wallets with different descriptors.
```suggestion
        let (wallet1, txid1) = get_funded_wallet_single(get_test_wpkh());
```